### ISSUE TITLE
[feature] 'Strict mode' for =~ operator, requiring regexp on RHS

### ISSUE THREAD

**Issue by epa** [Perl/perl5#18355]
The `=~` operator normally takes a `m//`, `tr///`, or `s///` operator on the right hand side.  However, if the RHS is an expression, it's taken as a regexp and the `m//` operation is implied.  This can be useful, but I wish for greater strictness.  I would like to require the RHS to be a compiled regexp object, if it's not one of the three operators.

```
my $re = qr/bar/;
my $fred = 'jim';

$x =~ /foo/;    # fine, regexp known at compile time
$x =~ /$fred/;  # fine, regexp built at run time
$x =~ $re;      # fine, it's a compiled regexp object
$x =~ $fred;    # I would like this to fail at run time with 'RHS of =~ is not a compiled regexp'
```

Taking the RHS as an implicit regexp match is longstanding behaviour and can't really be changed as a default, but I would like an option for the stricter behaviour.  For me, this would help with some typos, and with cases where I had intended to compile my regexps with `qr//` for speed but some got missed out.  It's usually quite straightforward to add the explicit `//` or `qr//` so I don't feel that Perl would become less expressive with the strictness on.

**Comment by Grinnz** [comment]
This would be a good lexical feature IMO. It could have both compile time (for literals) and runtime (for variables that may contain regexp ref) checks. It could even be entirely compile-time and require that you add a `m//` operator for regexp refs (which IIRC results in the same operations).

**Comment by Grinnz** [comment]
One other consideration is objects with `qr` overload, either direct or generated from another overload. IMO these should be treated the same as regexp refs.

**Comment by epa** [comment]
Agreed.  It's really only the feature of promoting an ordinary scalar into a regular expression match that I'd like to suppress.  I would be very happy if it gave a warning, which I could then fatalize.  That's analogous to other implicit conversions which have been there since perl 1.0, but nowadays have warnings you can enable and most people turn on.  However, to include the new warning as part of the standard `use warnings` set might be a step too far.

**Comment by Grinnz** [comment]
New warnings are acceptable in cases where it prevents mistakes more often than providing annoyance - this right is explicitly reserved in warnings.pm. I think the more relaxed interpretation of this (compile check for literals, runtime check otherwise) would fall under that interpretation.

**Comment by Grinnz** [comment]
On the other hand, auto-conversion of strings to regex matches are relied upon quite a lot so I would prefer to see it as a lexical feature that has to be requested.
